hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieBaseFileGroupRecordBuffer.java

```diff
@@ -237,7 +241,8 @@ protected Option<Pair<T, Map<String, Object>>> doProcessNextDataRecord(T record,
             Comparable incomingOrderingValue = readerContext.getOrderingValue(
                 Option.of(record), metadata, readerSchema, orderingFieldName, orderingFieldTypeOpt, orderingFieldDefault);
             if (incomingOrderingValue.compareTo(existingOrderingValue) > 0) {
-              return Option.of(Pair.of(record, metadata));
+              return Option.of(Pair.of(isDeleteRecord(Option.of(record), (Schema) metadata.get(INTERNAL_META_SCHEMA))
```

Comment:
I feel we should not have 2 ways of handling deletes within doProcessNextDataRecord. may be, we should process records from data block, and call into processNextDeletedRecord() if we detect a deleted record. 

that way, we have only one way of handling a deleted record. 

and processNextDataRecord will only take in valid records. we can avoid calling isDeleteRecord(Option.of(record)) within processNextDataRecord.

Reply:
but lets try to land this patch. and do an immediate follow up.